compiler/rustc_codegen_ssa/src/mir/intrinsic.rs

```diff
@@ -462,16 +480,29 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                     _ => unreachable!(),
                 };
 
-                let ty = fn_args.type_at(0);
-                if int_type_width_signed(ty, bx.tcx()).is_some() || ty.is_raw_ptr() {
-                    let ordering = fn_args.const_at(1).to_value();
-                    let ptr = args[0].immediate();
-                    let val = args[1].immediate();
-                    bx.atomic_rmw(atom_op, ptr, val, parse_atomic_ordering(ordering))
-                } else {
-                    invalid_monomorphization_int_type(ty);
-                    return Ok(());
-                }
+                // The type of the in-memory data.
+                let ty_mem = fn_args.type_at(0);
+                // The type of the 2nd operand, given by-value.
+                let ty_op = fn_args.type_at(1);
+
+                let ordering = fn_args.const_at(2).to_value();
+                let ptr = args[0].immediate(); // of type "pointer to `ty_mem`"
+                let val = args[1].immediate(); // of type `ty_op`
+                // We require either both arguments to have the same integer type, or the first to
+                // be a pointer and the second to be `usize`.
+                let val_adjusted =
+                    if int_type_width_signed(ty_mem, bx.tcx()).is_some() && ty_op == ty_mem {
+                        val
+                    } else if ty_mem.is_raw_ptr() && ty_op == bx.tcx().types.usize {
+                        // FIXME: LLVM does not support an atomic "add integer to pointer" operation, so
+                        // we need to instead add two pointers and hope that works out. See
+                        // <https://github.com/llvm/llvm-project/issues/120837>.
+                        bx.inttoptr(val, bx.backend_type(bx.layout_of(ty_mem)))
```

Comment:
This is now doing something pretty weird where this code casts the integer to pointer and then bx.atomic_rmw is going to cast it back to integer.

LLVM already expects the arguments to atomicrmw here to be a pointer and an integer. Now that the Rust intrinsic no longer artificially forces both arguments to be pointers, we should remove both this cast and the argument cast in atomic_rmw.

The only cast you need to keep is the cast of the return value of atomicrmw.